Session8 状態管理を見直す

yumemi-rnose opened the pull request:
## 課題

close #9 

## 対応箇所

- [x] riverpod導入
- [x] riverpodで状態管理
- [x] [アーキテクチャ導入](https://qiita.com/cocoa-maemae/items/e3f2eabbe0877c2af8d0)
- [x] riverpod_graphを使ってARCHITECTURE.mdファイル作成

## 動作

※ 動作に変更ないので割愛させていただきます

yumemi-rnose committed:
feat: Riverpod導入

yumemi-rnose committed:
feat: onion architecture採用し、viewStateをriverpod で管理

https://qiita.com/cocoa-maemae/items/e3f2eabbe0877c2af8d0

yumemi-rnose committed:
feat: riverpod_graphを使ってARCHITECTUREファイルを出力

yumemi-rnose committed:
fix: build 関数直下での ref.read

nerd0geek1 commented on lib/application/weather_service.dart:
```diff
@@ -0,0 +1,24 @@
+WeatherService weatherService(WeatherServiceRef ref) =>
+    WeatherService(ref.watch(fetchWeatherMixinProvider));
+
+mixin FetchWeatherMixin {
```
@yumemi-rnose 
こちら、どういった意図をもって宣言したものですかね？

Swiftのprotocolのように型で縛り、実体については各個で実装するように、という形ですかね？

であれば、Dartについては`Implicit interfaces`があるため、この部分がまるっと不要になるかと
思いましたが、いかがでしょう？
https://dart.dev/language/classes#implicit-interfaces

yumemi-rnose replied:
@nerd0geek1 
> こちら、どういった意図をもって宣言したものですかね？

こちらはRepositoryのinterfaceですね。

yumemi-rnose replied:
@nerd0geek1 
> Swiftのprotocolのように型で縛り、実体については各個で実装するように、という形ですかね？
> 
> であれば、DartについてはImplicit interfacesがあるため、この部分がまるっと不要になるかと
思いましたが、いかがでしょう？
https://dart.dev/language/classes#implicit-interfaces

すみません。Implicit interfacesをどこにどう使ったらいいかわからないです。

blendthink replied:
@yumemi-rnose 
横から失礼します、、！（ゆめみの岡山です）
Dart では、すべてのクラスは、そのクラスとそのクラスが実装するインターフェースのすべてのインスタンス・メンバーを含むインターフェースを暗黙的に定義されるので、わざわざインターフェースを用意しなくてもテストでインスタンスを差し替えることが簡単にできます。
そのため、テストしやすくする目的で `FetchWeatherMixin` を `Repository` の `interface` として用意したのであれば `FetchWeatherMixin` を削除して、 `WeatherService` のコンストラクタに `WeatherRepository` を直接渡すようにすれば良さそうです。

```dart
import 'package:flutter_training/domain/weather.dart';
import 'package:flutter_training/repository/weather_repository.dart';

class WeatherService {
  WeatherService(this._repository);

  final WeatherRepository _repository;

  Weather fetchWeather() {
    return _repository.execute();
  }
}
```

yumemi-rnose replied:
@blendthink 
レビュー、 `Implicit interfaces` の解説ありがとうございます。
テストしやすくする為、というよりはアーキテクチャに基づき下位層が上位層に依存しないような意味で `Mixin`という `Repository`の `interface` を用意したっていう意図でした。確かに暗黙的に `interface` が定義されるのであればわざわざ定義しなおさなくてもいいのかもですね。
ですが、 `WeatherRepository` が違うものになった場合でも `WeatherService` にはimport文含め影響うけないようにしたいです。DIしているとこはさすがに影響しますが。

blendthink replied:
@yumemi-rnose 
ご返信ありがとうございます！

> ですが、 WeatherRepository が違うものになった場合でも WeatherService にはimport文含め影響うけないようにしたいです。DIしているとこはさすがに影響しますが。

このような理由があるのであれば、個人的にはいまの実装のままでも問題ないかなと思います 🙆‍♂️

nerd0geek1 commented:
`ui/`ディレクトリについて、研修では良いですが、実際のプロジェクトではscreenごとにディレクトリ切ったほうが良いですね。

目にすることが多いのは以下のような構成です。
- `ui | presentation`
  - `screens/`
    - `green_back/`
    - `weather/`
      - `weather_screen.dart`
      - `weather_screen_state.dart`
  - `components/`

yumemi-rnose committed:
fix: ui内のフォルダ構成をscreen毎にする

nerd0geek1 commented:
(今回は修正してもしなくても良いです)
また、個人的な感覚になりますが、`XXXWidget`系のクラスにおいては、buildメソッドを各propertyの次に書いたほうが読みやすさが上がるように感じました。

理由としては以下の通りです。
- すべてのWidgetファイルはbuildメソッドを持つため、すべてのdartファイルにおいて場所が同じだと読みやすい
  - buildメソッドの上にprivateメソッドがあると、画面によって位置がズレるため、斜め読みしにくい。
- buildファイルの中身 = iOSのxibファイル、のように画面レイアウトそのものであるため、Widgetファイルを見た際に
優先度高く見たいものである


また、`WeatherScreen`の中身で内容の重複が大きい部分については、privateなWidgetとして宣言し利用するようにすると、分量が減り、読みやすくなるように感じました。(以下のような部分ですね)
```
Expanded(
                        child: Padding(
                          padding: const EdgeInsets.symmetric(vertical: 16),
                          child: Text(
                            '${state?.minTemperature.toString() ?? '**'} ℃',
                            style: textStyle?.copyWith(color: Colors.blue),
                            textAlign: TextAlign.center,
                          ),
                        ),
                      ),
                      Expanded(
                        child: Padding(
                          padding: const EdgeInsets.symmetric(vertical: 16),
                          child: Text(
                            '${state?.maxTemperature.toString() ?? '**'} ℃',
                            style: textStyle?.copyWith(color: Colors.red),
                            textAlign: TextAlign.center,
                          ),
                        ),
                      )
```

yumemi-rnose committed:
fix: DIと実装を分けた

yumemi-rnose committed:
fix: 画面のComponent化

yumemi-rnose commented:
> (今回は修正してもしなくても良いです) また、個人的な感覚になりますが、`XXXWidget`系のクラスにおいては、buildメソッドを各propertyの次に書いたほうが読みやすさが上がるように感じました。
> 
> 理由としては以下の通りです。
> 
> * すべてのWidgetファイルはbuildメソッドを持つため、すべてのdartファイルにおいて場所が同じだと読みやすい
>   
>   * buildメソッドの上にprivateメソッドがあると、画面によって位置がズレるため、斜め読みしにくい。
> * buildファイルの中身 = iOSのxibファイル、のように画面レイアウトそのものであるため、Widgetファイルを見た際に
>   優先度高く見たいものである
> 
> また、`WeatherScreen`の中身で内容の重複が大きい部分については、privateなWidgetとして宣言し利用するようにすると、分量が減り、読みやすくなるように感じました。(以下のような部分ですね)
> 
> ```
> Expanded(
>                         child: Padding(
>                           padding: const EdgeInsets.symmetric(vertical: 16),
>                           child: Text(
>                             '${state?.minTemperature.toString() ?? '**'} ℃',
>                             style: textStyle?.copyWith(color: Colors.blue),
>                             textAlign: TextAlign.center,
>                           ),
>                         ),
>                       ),
>                       Expanded(
>                         child: Padding(
>                           padding: const EdgeInsets.symmetric(vertical: 16),
>                           child: Text(
>                             '${state?.maxTemperature.toString() ?? '**'} ℃',
>                             style: textStyle?.copyWith(color: Colors.red),
>                             textAlign: TextAlign.center,
>                           ),
>                         ),
>                       )
> ```

[49900d4](https://github.com/yumemi-rnose/yumemi-flutter-training/pull/21/commits/49900d4ec6f7a12c6b95c67eb14ad71defe6c7b3) にて修正しました。
認識合ってますでしょうか？

nerd0geek1 commented:
@yumemi-rnose 
https://github.com/yumemi-rnose/yumemi-flutter-training/pull/21#issuecomment-2258061554
はい！方向性としてはあってます！
(コードが重複していないコンポーネントについては、独立したWidgetとして定義しなくても問題ないとは思いましたが)

nerd0geek1 commented on lib/ui/screen/weather/weather_screen.dart:
```diff
@@ -0,0 +1,199 @@
+      : _text = text,
+        _onPressed = onPressed;
+
+  final String _text;
+  final Function _onPressed;
```
@yumemi-rnose 
ちなみに、Widget全般についてですが、引数はprivateなpropertyとしなくても問題ありません。
(UIKitなどと異なり宣言的UIであるため、後からアクセスしにくく、publicでも問題になることがあまりない)

nerd0geek1 replied:
@yumemi-rnose 
それ以外でいうと、`onPress`は`Function`で宣言するより[VoidCallback](https://api.flutter.dev/flutter/dart-ui/VoidCallback.html)を使うのが良いかと思います。
([`TextButton`](https://api.flutter.dev/flutter/material/TextButton-class.html)も`VoidCallback`を引数としていますし、また現状正常に動作していないと思います)

yumemi-rnose replied:
@nerd0geek1 
> また現状正常に動作していないと思います

→ 失礼しました。[bf9f948](https://github.com/yumemi-rnose/yumemi-flutter-training/pull/21/commits/bf9f9488e70ac7e0fa85ca82f7a9b05415c669ff) にて修正しました。

yumemi-rnose committed:
fix: クリックイベントが動作していなかった

nerd0geek1 commented on lib/ui/screen/weather/weather_screen.dart:
```diff
@@ -0,0 +1,201 @@
+    final textStyle = Theme.of(context).textTheme.labelLarge;
+    return Expanded(
+      child: TextButton(
+        onPressed: () {
+          _onPressed?.call();
+        },
```
@yumemi-rnose 
こちらですが、`TextButton.onPressed`も引数の`onPressed`もどちらも型が`VoidCallback`なので
`onPressed: _onPressed`と書けます。

yumemi-rnose replied:
@nerd0geek1 
ありがとうございます。
[65b0455](https://github.com/yumemi-rnose/yumemi-flutter-training/pull/21/commits/65b04556e65889d61dda8750fbf1525f6df76acd) 修正しました。

nerd0geek1 commented on lib/ui/screen/weather/weather_screen.dart:
```diff
@@ -0,0 +1,201 @@
+
+  @override
+  Widget build(BuildContext context, WidgetRef ref) {
+    final notifier = ref.watch(weatherScreenStateProvider.notifier);
```
@yumemi-rnose 
`ref.watch`は値が変化したタイミングでrebuildが必要な場合に使います。

したがって、76行目に移動し、`ref.read`とするのが正しいです。
**公式のドキュメント**
https://riverpod.dev/docs/concepts/reading#using-ref-to-interact-with-providers

また、内部の動きについては以下が詳しいです。
https://zenn.dev/chooyan/books/92a0a489f68233/viewer/watch

blendthink replied:
@nerd0geek1 @yumemi-rnose 
ドキュメントにも記載されていますが Notifier 内の State が更新されても rebuild は走らないようになっているため、現在の実装でも問題なく正しいです。実際に `build` メソッド内でブレイクポイントを貼って確かめてみると分かります。

![image](https://github.com/user-attachments/assets/7a947a5c-4089-43bc-89ac-f7082aa07843)

プロジェクトによってはこのような場合に、 `ref.read` を使うように統一していることがあるため、その場合は方針に従うでいいかと思います。あとは、問題ないことが分かりやすくなるように `ref.read` を使用するとしてもいいかなとは思います！

`ref.read` を使うようにすると記述量がどうしても多くなってしまうことがあるため、別案件では `build` メソッド内で Notifier を `ref.watch` する書き方を許容していたりします💡

yumemi-rnose replied:
@nerd0geek1 
[2beff93](https://github.com/yumemi-rnose/yumemi-flutter-training/pull/21/commits/2beff936907d8133e1c77919d0a222d7cdb32339) 修正してみました。認識あってますでしょうか？

yumemi-rnose committed:
fix: _onPressedを引数にそのまま指定

yumemi-rnose committed:
fix: クリックイベント時はref.readを使用する

https://riverpod.dev/docs/concepts/reading#using-ref-to-interact-with-providers

nerd0geek1 reviewed (approved):
数々の対応ありがとうございました！
LGTMです！
![LGTM](https://i.lgtm.fun/2t5e.png)

yumemi-rnose committed:
update: ARCHITECTURE.md
